052-record-reconciled-versions.md

```diff
@@ -0,0 +1,185 @@
+- StrimziPodSet reconciler reconciles the `SPS`s creating the Kafka Brokers
+- Kafka Reconciler picks up that the Kafka Brokers managed by the SPS are now Ready and finished installing.
+  Updating the status with the reconciled kafka versions:
+  ```
+  status:
+    versions:
+      kafka: 3.5.1
+  ```
+- Kafka Reconciler continues to install other components such as entity operator.
+  Once all components are deployed and updated, update `Kafka` CR with
+  ```
+  status:
+    versions:
+      lastSuccessfulReconciliationBy: 0.38.0
+      kafka: 3.5.1
+  ```
+- User can now verify `Kafka.status.versions.lastSuccessfulReconciliationBy` and `Kafka.status.versions.kafka` to check if it has reconciled to correct/expected levels.
+- if there is an error at any point prior to the Kafka Broker rollout `Kafka.status.versions.kafka` will not be updated.
+- if there is an error at any point during reconcile `Kafka.status.versions.lastSuccessfulReconciliationBy` will not be updated.
+
+
+### Upgrade (with new mechansim)
+
+This defines an example where it is imagined that `0.37.0` had this mechanism already implemented, and is now upgrading to a new version `0.38.0` (with kafka upgrading from `3.5.0` to `3.5.1`)
+
+- On an upgrade, a user has a Kafka CR:
+  ```
+  status:
+    versions:
+      lastSuccessfulReconciliationBy: 0.37.0
+      kafka: 3.5.0
+  ```
+- The KafkaReconciler reconciles the CR and creates `StrimziPodSet` CRs
+- StrimziPodSet reconciler reconciles the `SPS`s creating the Kafka Brokers
+- Kafka Reconciler picks up that the Kafka Brokers managed by the SPS are now Ready and finished installing.
+  Updating the status with the reconciled kafka versions:
+  ```
+  status:
+    versions:
+      lastSuccessfulReconciliationBy: 0.37.0
+      kafka: 3.5.1
+  ```
+- Kafka Reconciler continues to install other components such as entity operator.
+  Once all components are deployed and updated, update `Kafka` CR with
+  ```
+  status:
+    versions:
+      lastSuccessfulReconciliationBy: 0.38.0
+      kafka: 3.5.1
+  ```
+- User can now verify `Kafka.status.versions.lastSuccessfulReconciliationBy` and `Kafka.status.versions.kafka` to check if it has reconciled to correct/expected levels.
+- if there is an error at any point prior to the Kafka Broker rollout `Kafka.status.versions.kafka` will not be updated.
+- if there is an error at any point during reconcile `Kafka.status.versions.lastSuccessfulReconciliationBy` will not be updated.
+
+## `KafkaNodePool` considerations
+
+Some users use new `KafkaNodePool` CRs to facilitate upgrades through rolling out new brokers with new Kafka versions, reassigning the partitions and then delete the old brokers.
+This proposal should work with this method od upgrade, since the Kafka version + readiness checks will be track by the `StrimziPodSet`, so that when a user had finished upgrading via this method, the new `status.versions.kafka` should be updated accordingly.
```

Comment:
I think the future consideration to make here is what if different node pools have different Kafka versions. And thus there will be no single version for the `status.versions.kafka` field (as multiple of them will be used). And this is not answered in this paragraph.

_Note, we might never support something like this in reality. But I expect that this will be raised by some users in the future._